crates/ruff_linter/src/rules/refurb/rules/slice_to_remove_prefix_or_suffix.rs

```diff
@@ -0,0 +1,498 @@
+    let slice = slice.as_slice_expr()?;
+    let test_name = &test
+        .as_call_expr()?
+        .func
+        .as_attribute_expr()?
+        .value
+        .as_name_expr()?
+        .id;
+    let func_name = test
+        .as_call_expr()?
+        .func
+        .as_attribute_expr()?
+        .attr
+        .id
+        .as_str();
+
+    let func_args = &test.as_call_expr()?.arguments.args;
+    if func_args.len() != 1 {
+        return None;
+    }
+
+    let affix = &func_args[0];
+    match (body_name, test_name, else_or_target_name, slice, func_name) {
+        (text, a, b, _, "startswith") if text == a && a == b && slice.upper.is_none() => {
+            Some(RemoveAffixData {
+                text,
+                bound: slice.lower.as_ref()?,
+                affix_query: AffixQuery {
+                    kind: AffixKind::StartsWith,
+                    affix,
+                },
+            })
+        }
+        (text, a, b, _, "endswith") if text == a && a == b && slice.lower.is_none() => {
+            Some(RemoveAffixData {
+                text,
+                bound: slice.upper.as_ref()?,
+                affix_query: AffixQuery {
+                    kind: AffixKind::EndsWith,
+                    affix,
+                },
+            })
+        }
+        _ => None,
+    }
+}
+
+/// Tests whether the slice of the given string actually removes the
+/// detected affix.
+///
+/// For example, in the situation
+///
+/// ```python
+///  text[:bound] if text.endswith(suffix) else text
+/// ```
+///
+/// this function verifies that `bound == -len(suffix)` in two cases:
+///     - `suffix` is a string literal and `bound` is a number literal
+///     - `suffix` is an expression and `bound` is
+///     exactly `-len(suffix)` (as AST nodes, prior to evaluation.)
+fn affix_matches_slice_bound(data: &RemoveAffixData) -> bool {
+    let RemoveAffixData {
+        text: _,
+        bound,
+        affix_query: AffixQuery { kind, affix },
+    } = *data;
+
+    match (kind, bound, affix) {
+        (
+            AffixKind::StartsWith,
+            ast::Expr::NumberLiteral(ast::ExprNumberLiteral {
+                value: num,
+                range: _,
+            }),
+            ast::Expr::StringLiteral(ast::ExprStringLiteral {
+                range: _,
+                value: string_val,
+            }),
+        ) => num.as_int().is_some_and(|x| {
+            let length = u32::try_from(string_val.len())
+                // Only support prefix removal for size at most `u32::MAX`.
+                .unwrap_or(u32::MAX);
+            x == &ast::Int::from(length)
+        }),
+        (
+            AffixKind::StartsWith,
+            ast::Expr::Call(ast::ExprCall {
+                range: _,
+                func,
+                arguments,
+            }),
+            _,
+        ) => {
+            func.as_name_expr()
+                .is_some_and(|name| name.id == ast::name::Name::new("len"))
```

Comment:
This won't work if somebody is annoying and does something like this:

```py
from builtins import len as builtins_len
```

You want to use something like `checker.semantic().match_builtin_expr(func, "len")`, which will recognise even `builtins_len` as a reference to the `builtins.len()` function

https://github.com/astral-sh/ruff/blob/a4ebe7d34407ea353762ebde1fef4a718edddb55/crates/ruff_python_semantic/src/model.rs#L308-L326

Reply:
Nice! Done.